File: python/metaparticle/runner/docker.py
import os


def ports(ports):
    try:
        port_string = ""
        for port in ports:
            port_string += " -p {port}:{port}".format(port=port)
        return port_string
    except KeyError:
        return ' '


def run(img, name, options):
    # Launch docker container
    os.system('docker run --rm --name {} {} -d {}'.format(
        name,
        ports(options.ports),
        img
    ))
    # Attach to logs
    os.system('docker logs -f {}'.format(name))


def cancel(name):
    os.system('docker kill {}'.format(name))
    os.system('docker rm {}'.format(name))

File: rust/src/executor/docker.rs
use super::run_docker_process;
use super::Runtime;
use Executor;

pub struct DockerExecutor{}

impl Executor for DockerExecutor {
    fn cancel(&self, name: &str) {
        run_docker_process(vec!["docker", "stop", name]);
        run_docker_process(vec!["docker", "rm", "-f", name]);
    }

    fn logs(&self, name: &str) {
        run_docker_process(vec!["docker", "logs", "-f", name]);
    }

    fn run(&self, image: &str, name: &str, config: Runtime) {
        let mut ports = String::new();
        let mut args = vec!["docker", "run", "-d", "--rm", "--name", name];
        
        if let Some(port) = config.ports {
            ports.push_str(&format!("-p {port}", port=port));
            args.push(&ports);
        }

        args.push(image);

        run_docker_process(args);
    }
}

